From 4511edb4a14f39fbcb5bc0d178f4c91b2e88a04c Mon Sep 17 00:00:00 2001
From: gamingdoom <37276884+gamingdoom@users.noreply.github.com>
Date: Sat, 30 Dec 2023 16:10:12 -0800
Subject: [PATCH] More robust BeamNG.Drive VR Patch

---
 openxr.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/openxr.c b/openxr.c
index 0a4b9c18..2f6af119 100644
--- a/openxr.c
+++ b/openxr.c
@@ -1600,14 +1600,22 @@ XrResult WINAPI wine_xrGetVulkanDeviceExtensionsKHR(XrInstance instance, XrSyste
 
     WINE_TRACE("%p, 0x%s, %u, %p, %p\n", instance, wine_dbgstr_longlong(systemId), bufferCapacityInput, bufferCountOutput, buffer);
 
-    if(bufferCapacityInput == 0){
+    const char *sgi = getenv("SteamGameId");
+
+    if(bufferCapacityInput == 0 && ((!sgi) | (sgi && strcmp(sgi, "284160") != 0))){
         *bufferCountOutput = sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME);
         return XR_SUCCESS;
+    } else if (bufferCapacityInput == 0 && (sgi && strcmp(sgi, "284160") == 0)){
+        *bufferCountOutput = 0;
+        return XR_SUCCESS;
     }
 
-    if(bufferCapacityInput < sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME)){
+    if((bufferCapacityInput < sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME) && ((!sgi) | (sgi && strcmp(sgi, "284160") != 0)))){
         *bufferCountOutput = sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME);
         return XR_ERROR_SIZE_INSUFFICIENT;
+    } else if (bufferCapacityInput < sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME) && (sgi && strcmp(sgi, "284160") == 0)){
+        *bufferCountOutput = 0;
+        return XR_SUCCESS;
     }
 
     res = ((wine_XrInstance *)instance)->funcs.p_xrGetVulkanDeviceExtensionsKHR(
@@ -1634,8 +1642,10 @@ XrResult WINAPI wine_xrGetVulkanDeviceExtensionsKHR(XrInstance instance, XrSyste
 
     heap_free(buf);
 
-    memcpy(buffer, WINE_VULKAN_DEVICE_EXTENSION_NAME, sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME));
-    *bufferCountOutput = sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME);
+    if ((!sgi) | (sgi && strcmp(sgi, "284160") != 0)) {
+        memcpy(buffer, WINE_VULKAN_DEVICE_EXTENSION_NAME, sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME));
+        *bufferCountOutput = sizeof(WINE_VULKAN_DEVICE_EXTENSION_NAME);
+    }
 
     return XR_SUCCESS;
 }
-- 
2.43.0

